pandas/core/frame.py

```diff
@@ -3733,6 +3733,18 @@ def _set_item(self, key, value) -> None:
                 if isinstance(existing_piece, DataFrame):
                     value = np.tile(value, (len(existing_piece.columns), 1)).T
 
+        if (
+            value.ndim == 2
+            and value.shape[1] > 1
+            and not isinstance(self.columns, MultiIndex)
+            and len(self.columns) == len(set(self.columns))
```

Comment:
You can just use self.columns.is_unique

This would not catch something like column is duplicated 2 times but you got 3 columns in your numpy array

Reply:
get_indexer_for would help here probably